Continued API update (and other changes)

bottleofench opened the pull request:
Still in development!
If you want to test this updates, here is the [updated jar](https://cdn.discordapp.com/attachments/760175424620986388/1004628343645282335/plasmovoice-server-1.0.12.jar)
(The link will be updated after almost every commit)

**Referring to third-party developers:**
Write what else you would like to see in the API.

API changes:
- New API methods: 
1) `isTalking()` ; 
2) `getMutedPlayersUUIDs()` ; 
3) `getMuteReason()` ; 
4) `getTimestampOfMuteEnd()` ;
5) `getMaxDistance()` ;
6) `getDefaultDistance()` ;
7) `getMaxPriorityDistance`
- Added Javadocs for all events;
- Updated PlayerVoiceMuteEvent and PlayerConfigEvent;
- Added `@NotNull` annotations to `sendVoicePacketToPlayer()` method;
- Removed `@ApiStatus.Experimental` annotations for API methods;
- Remove unneccesary import in PlayerVoiceUnmuteEvent;
- Used `@Getter` in all event classes;
- Updated API examples in README.md;
- Renamed `getPlayers()` method to `getConnectedPlayersUUIDs()` ;
- Fixed javadoc for `getPlayerModLoader()` method;
- Added `!recipient.isOnline` check to `sendVoicePacketToPlayer()` method;
- Replaced Packet to VoiceServerPacket in PlayerSpeakEvent

Other changes:
- Added missing `@NotNull` annotations in commands classes;
- Replaced statement lambda to expression lambda in PlayerListener;
- Modified regex for VoiceMute class (`[0-9] -> \\d`);
- Replaced `size() == 0` checks to `isEmpty()` ;
- Merged two if-statements in one in main class;
- Replaced deprecated `{version}` to `{project.version}` in pom.xml;
- Used Collections.emptyList() instead of null;
- Merged some if-statements;
- Optimized PlaceholderPlasmoVoice class;
- Replaced `player.getFirstPlayed() == 0` check to `!player.hasPlayedBefore()` in VoiceUnmute class;
- Added `disableLogs` to PlasmoVoiceConfig class;
- Remove nested try-catch block in SocketServerUDPQueue;
- Added `@Override` annotations in SocketServerUDP and SocketServerUDPQueue

bottleofench committed:
Continued API update (and some other changes).

API changes:
- New API methods:
  1) isTalking();
- Added Javadocs for all events;
- Updated PlayerVoiceMuteEvent;
- Added @NotNull annotations to sendVoicePacketToPlayer() method;
- Removed @ApiStatus.Experimental annotations for API methods;
- Optimized imports in PlayerVoiceUnmuteEvent;
- Used lombok.Getter in all event classes;
- Updated API examples in README.md;
- Renamed getPlayers() method to getConnectedPlayersUUIDs();
- Fixed javadoc for getPlayerModLoader() method;
- Added !recipient.isOnline check to sendVoicePacketToPlayer() method.

Other tiny changes:
- Added missing @NotNull annotations in command classes;
- Replaced statement lambda to expression lambda in PlayerListener;
- Modified regex for VoiceMute class ([0-9] -> \\d);
- Replaced size() == 0 checks to isEmpty();
- Merged two if-statements in one in main class;
- Replaced deprecated {version} to {project.version} in pom.xml.

bottleofench committed:
Fix README.md

bottleofench commented:
Oh, I was just doing a remote voice chat between players, and I generally succeeded, except with a very questionable solution. 

Here's the event handler code:
```java
@EventHandler
private void onSpeak(PlayerSpeakEvent event) {
    if (isHeadphonesMuted(event.getPlayer().getName())) return;
    if (isMicrophoneMuted(event.getPlayer().getName())) return;
    List<String> list = getGroups(event.getPlayer().getName());
    if (list.isEmpty()) return;
    for (String group : list) {
        for (String playerName : getPlayers(group)) {
            if (event.getPlayer().getName().equals(playerName)) continue;
            Player player = Bukkit.getPlayerExact(playerName);
            if (player == null || !player.isOnline()) continue;
            if (isHeadphonesMuted(playerName)) continue;
            VoiceServerPacket pa = (VoiceServerPacket) event.getPacket();
            VoiceServerPacket packet = new VoiceServerPacket(pa.getData(), player.getUniqueId(), pa.getSequenceNumber(), pa.getDistance());
            api.sendVoicePacketToPlayer(packet, player);
        }
    }
}
```

I create a separate packet with the same data, and change the original UUID to the UUID of the player the packet will be sent to. And, it generally works! The only question is, would it be right to do it that way? Is there another solution to this problem?

bottleofench committed:
Update README.md

apehum commented:
> Oh, I was just doing a remote voice chat between players, and I generally succeeded, except with a very questionable solution.
> 
> Here's the event handler code:
> 
> ```java
> @EventHandler
> private void onSpeak(PlayerSpeakEvent event) {
>     if (isHeadphonesMuted(event.getPlayer().getName())) return;
>     if (isMicrophoneMuted(event.getPlayer().getName())) return;
>     List<String> list = getGroups(event.getPlayer().getName());
>     if (list.isEmpty()) return;
>     for (String group : list) {
>         for (String playerName : getPlayers(group)) {
>             if (event.getPlayer().getName().equals(playerName)) continue;
>             Player player = Bukkit.getPlayerExact(playerName);
>             if (player == null || !player.isOnline()) continue;
>             if (isHeadphonesMuted(playerName)) continue;
>             VoiceServerPacket pa = (VoiceServerPacket) event.getPacket();
>             VoiceServerPacket packet = new VoiceServerPacket(pa.getData(), player.getUniqueId(), pa.getSequenceNumber(), pa.getDistance());
>             api.sendVoicePacketToPlayer(packet, player);
>         }
>     }
> }
> ```
> 
> I create a separate packet with the same data, and change the original UUID to the UUID of the player the packet will be sent to. And, it generally works! The only question is, would it be right to do it that way? Is there another solution to this problem?

Currently there is no way to send packets directly to the player without any player "source". So, probably, your solution is only way to do it.

bottleofench commented:
Okay, thanks for the answer! I have already updated a first example with this in the README.md.

bottleofench committed:
- New API method: `getMutedPlayersUUIDs()`;
- Updated README.md with a bunch of API usage examples.

bottleofench committed:
- New API methods: `getMuteReason()` and `getTimestampOfMuteEnd()`

bottleofench commented:
I have a few questions about the code:
1) What if instead of returning strings `forge` or `fabric` in method `getPlayerModLoader()` it returns enum like `ModLoader.FORGE` or `ModLoader.FABRIC`? Wouldn't that be much more convenient than constantly initializing the strings?
2) Also, I think it would be nice to make a PlasmoVoiceMessages class, similar to PlasmoVoiceConfig, which would store all messages from config. Or just store these messages in the same PlasmoVoiceConfig. This would be convenient and much easier for the server to process.

Also I would like to see at least a simplified version of [this](https://github.com/U61vashka/PlasmoVoiceAddon) in the plugin. I think it could be a very good part of the API. What do you think about it? I might try this to add it to the plugin and create a PR for it.

bottleofench committed:
1) New API methods:
- `getMaxDistance()`;
- `getDefaultDistance()`;
- `getMaxPriorityDistance`
2) Some tweaks:
- Use Collections.emptyList() instead of null;
- Merge some if-statements;
- Optimize PlaceholderPlasmoVoice class;

bottleofench committed:
Added `disableLogs` to PlasmoVoiceConfig class

bottleofench committed:
Replaced Packet to VoiceServerPacket in PlayerSpeakEvent

bottleofench committed:
Remove nested try-catch block in SocketServerUDPQueue

bottleofench committed:
Added `@Override` annotation in SocketServerUDPQueue

bottleofench committed:
Added `@Override` annotation in SocketServerUDP

bottleofench committed:
Merged two nested if-statements in SocketServerUDPQueue

bottleofench commented:
Added small TODO-comment in SocketServerUDPQueue. You can review that.

bottleofench committed:
Fixed /unmute command

bottleofench committed:
Used `@Getter` in PlayerVoiceUnmuteEvent

bottleofench committed:
Removed TODO-comment

bottleofench committed:
Added a javadoc for the PlasmoVoiceAPI class indicating my involvement in API development.

bottleofench committed:
Oooops

bottleofench commented:
> I have a few questions about the code:
> 
> 1. What if instead of returning strings `forge` or `fabric` in method `getPlayerModLoader()` it returns enum like `ModLoader.FORGE` or `ModLoader.FABRIC`? Wouldn't that be much more convenient than constantly initializing the strings?
> 2. Also, I think it would be nice to make a PlasmoVoiceMessages class, similar to PlasmoVoiceConfig, which would store all messages from config. Or just store these messages in the same PlasmoVoiceConfig. This would be convenient and much easier for the server to process.
> 
> Also I would like to see at least a simplified version of [this](https://github.com/U61vashka/PlasmoVoiceAddon) in the plugin. I think it could be a very good part of the API. What do you think about it? I might try this to add it to the plugin and create a PR for it.

Everything is ready to be merged at this point. I'll think more about this comment, and maybe create separate PRs for these changes.

bottleofench commented:
You can double-check everything before merging, but most likely everything should work fine.
